src/test/clojure/test/benchmark.clj

```diff
@@ -0,0 +1,150 @@
+    (ir->result-map f ir)))
+
+(defn benchmark-jarl [ir]
+  (benchmark-ir ir))
+
+(defn get-opa []
+  (get (System/getenv) "OPA" "opa"))
+
+(defn run-opa [policy query]
+  (let [result (shell/sh (get-opa) "eval" "--metrics" "-d" policy query)
+        err (:err result)]
+    (when (not (string/blank? err))
+      (throw (ex-info "Failed to run OPA eval" {:err err})))
+    (let [output (:out result)]
+      (json/read-str output))))
+
+(defn benchmark-opa [policy ir]
+  (let [f (fn [entry-point _]
+            (let [query (str "data." (string/replace entry-point #"/" "."))
+                  run (fn [] (let [result (run-opa policy query)
+                                   metrics (get result "metrics")
+                                   eval-ns (get metrics "timer_rego_query_eval_ns")
+                                   eval-micros (* eval-ns 0.001)]
+                               eval-micros))
+                  iterations 1000]
+              (->> (repeatedly iterations run)
+                   (reduce +)
+                   (* (/ 1 iterations)))))]
+    (ir->result-map f ir)))
+
+(defn benchmark-baseline []
+  (println "Measuring performance of baseline")
+  (let [bench-fn (fn [] (reduce + (range 100)))
+        bench-result (criterium/with-progress-reporting
+                       (criterium/benchmark* bench-fn {}))]
+    {"baseline" (mean bench-result)}))
+
+(defn csv-read [file]
+  (let [lines (with-open [reader (io/reader file)]
+                (reduce conj [] (line-seq reader)))
+        lines (map #(string/split % #",") lines)
+        categories (vec (first lines))
+        entries (next lines)
+        data (vec (map #(zipmap categories %) entries))]
+    {:categories (set categories)
+     :data       data}))
+
+(defn csv-write [csv file ordered-categories]
+  (let [categories (:categories csv)
+        data (:data csv)
+        ordered-categories (vec ordered-categories)         ; enforce insert order of into
+        unordered-categories (sort (filter #(not (some #{%} ordered-categories)) categories))
+        categories (into ordered-categories unordered-categories)
+        header (string/join "," categories)
+        text (loop [rows header
+                    data data]
+               (if (empty? data)
+                 rows
+                 (let [current-data (first data)
+                       row (string/join "," (map #(get current-data % "") categories))]
+                   (recur (str rows "\n" row) (next data)))))]
+    (spit file text)))
+
+(defn csv-add [csv entry]
+  (let [categories (union (:categories csv) (set (keys entry)))]
+    {:categories categories
+     :data       (conj (:data csv) entry)}))
+
+(defn git-revision []
+  (let [result (shell/sh "git" "rev-parse" "--short=10" "HEAD")
+        err (:err result)]
+    (when (not (string/blank? err))
+      (throw (ex-info "Failed to get OPA version" {:err err})))
+    (string/trim-newline (:out result))))
+
+(defn git-tag []
```

Comment:
Lots of things going on in this namespace, from benchmarking, CSV handling, git and running shell commands.. I suppose that's fine as long as we don't need any of that elsewhere, in which case we should move some shared stuff.

Reply:
Moved out git and csv stuff to their own namespaces.